### ISSUE TITLE
关于实验的一些疑问

### ISSUE THREAD

**Issue by peterjc123** [TsinghuaC3I/Intuitive-Fine-Tuning#2]
你好，感谢分享IFT部分的代码，这边做了一些实验，有一些疑问。
1. 首先我只用了embedding fusion那块，然后发现gsm8k和truthfulqa的效果有提升，其他的基本差不多
2. 然后我加上了dynamic relation propagation，发现有些指标有提升，但是gsm8k和mmlu都不太好
3. 我发现论文中学习率是5e-7，而我之前设置的是2e-5，进行了调整，然后发现效果更差了

我的疑问如下
1. 有没有embedding fusion这块的消融实验，单做这块理论上应该也是能说得通的吧
2. dynamic relation propagation这块对于我的数据集好像效果不太好，有什么建议吗？
3. 学习率这个有没有对照实验，我印象中ORPO论文中的学习率应该是接近SFT的，从你们论文中学习率应该是接近DPO的，但是试下来效果好像没有接近SFT的好？

**Comment by Hua-rookie** [comment]
【中文】
你好，我下面逐条回答你的问题：

1. Embedding Fusion：
    - 首先，这部分是提高因果性的核心算法之一，确实可以单独使用，所以gsm8k和truthfulqa这类强调因果性和事实性的benchmark会有提高（在你的实验、以及我们论文中的实验都有类似结论）；
    - 但是，用于Embedding Fusion的Temporary Residual Connection可能会导致误差过度累计（如果句长过长，或者模型当前能力过差），所以根据Bellman Equation的启示，需要进一步直接在Loss上引入“约束”来抑制这种误差累计，也就是Dynamic Relation Propagation；

2. Dynamic Relation Propagation：
    - 首先确认几个信息：① 我们使用的数据集是UltraFeedback_binarized；② 我们在关系传播时是将靠后token的loss像靠前token的位置累加；③ 累加系数这个变量需要参与梯度图的计算（因为关系累加也是优化目标的一部分，而不只是单纯的约束项）；④ 我们的实验是在 seq_length=1024 上进行，如果采用不同的seq_length，衰减系数 propagation_decay 可能需要适当调参；请确保上述实现一致，我们近期也会发布完整版的代码库供大家参考。。
    - 另外，mmlu这类多选题使用log_likelihood进行测试，和因果性生成任务间存在差距，所以我们的算法在这个benchmark上表现欠佳，具体原因也在论文的实验部分进行了分析。

3. Learning Rate：
    - 我们使用5e-7的学习率，其实是受到算力的限制。我们使用4张A6000-48GB显卡进行实验，所以只能设置 per_device_batch_size=1 or 2，对应的也就需要调小学习率来保证训练收敛；
    - 所以，我们设置的学习率只是提供同规模实验的复现参考，具体设置什么数值需要大家根据自己的实际情况调整。


【English】
Hello, below are my responses to your questions:

1. Embedding Fusion:
    - Firstly, this part is one of the core algorithms for enhancing causality and can indeed be used independently. Therefore, benchmarks like gsm8k and truthfulqa, which emphasize causality and factual accuracy, will see improvements (similar conclusions are drawn in your experiments and in experiments mentioned in our paper).
    - However, the Temporary Residual Connection used for Embedding Fusion may lead to excessive error accumulation (if the sentence length is too long or if the model's current capability is insufficient). Thus, inspired by the Bellman Equation, it's necessary to further introduce "constraints" directly on the loss to suppress such error accumulation, which is Dynamic Relation Propagation.

2. Dynamic Relation Propagation:
    - Please confirm some information: ① We used the UltraFeedback_binarized dataset. ② During relation propagation, we accumulate the loss from later tokens to earlier token positions. ③ The accumulation coefficient variable needs to participate in gradient calculation (because relation accumulation is also part of the optimization objective, not just a simple constraint). ④ Our experiments were conducted with seq_length=1024. If a different seq_length is used, the decay coefficient propagation_decay may need to be appropriately adjusted. Please ensure consistency in the above implementation. We will also release a complete code repository for reference soon.
    - Additionally, benchmarks like mmlu, which involve multiple-choice questions, are tested using log_likelihood, which differs from causal generation tasks. Therefore, our algorithm performs suboptimally on this benchmark, and the specific reasons are also analyzed in the experimental section of the paper.

3. Learning Rate:
    - We used a learning rate of 5e-7, mainly due to computational limitations. We conducted experiments using 4 A6000-48GB GPUs, so we could only set per_device_batch_size=1 or 2. Consequently, we needed to reduce the learning rate to ensure training convergence.
    - Therefore, the learning rate we set only serves as a reference for reproducing experiments of a similar scale. The specific value should be adjusted based on individual circumstances.

**Comment by peterjc123** [comment]
@Hua-rookie 感谢回复，还有个小问题，propagation_decay这个值我看论文里面用的是0.95，假设seq_length=1024，那么对于第一个token来说，他的loss的权重值就只有0.95**1024~=1.545154161087726e-23，这个是不是太小了呀？

**Comment by peterjc123** [comment]
感谢回复。关于Dynamic Relation Propagation这块应该是我们这边实现的问题，等后续修正之后如有新的结论会在这里同步的

**Comment by Hua-rookie** [comment]
【中文】
我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum

【English】
We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum

**Comment by peterjc123** [comment]
@Hua-rookie 不， 并不是weight计算这块的问题。准确的说是后面loss计算的逻辑，因为我们的数据并不像ultrafeedback那样都是比较长，会有一些比较短的数据。这里原来的代码是
```py
loss_ift = (-logps_further * cumsum_weight).sum(-1) / loss_mask.sum(-1)
return loss_ift.mean()
```
这样对于长度分布差距很大的情况，会使长序列的优化受到短序列的很大的影响。
我这边像改成下面这样
```py
loss_ift = (-logps_further * cumsum_weight).sum() / loss_mask.sum()
return loss_ift
```
这样貌似可以稍微抵消一下这些影响。你觉得如何？

**Comment by Hua-rookie** [comment]
> @Hua-rookie 不， 并不是weight计算这块的问题。准确的说是后面loss计算的逻辑，因为我们的数据并不像ultrafeedback那样都是比较长，会有一些比较短的数据。这里原来的代码是
> 
> ```python
> loss_ift = (-logps_further * cumsum_weight).sum(-1) / loss_mask.sum(-1)
> return loss_ift.mean()
> ```
> 
> 这样对于长度分布差距很大的情况，会使长序列的优化受到短序列的很大的影响。 我这边像改成下面这样
> 
> ```python
> loss_ift = (-logps_further * cumsum_weight).sum() / loss_mask.sum()
> return loss_ift
> ```
> 
> 这样貌似可以稍微抵消一下这些影响。你觉得如何？

【中文】
不好意思，我没有做过类似的实验，所以也说不好。
但是，ift在计算每条数据的loss时，相当于进行了长度上的normalization（除以loss_mask.sum(-1)就是在做这件事情）。在我理解里，这种normalization其实会适当减弱长度分布带来的影响？
所以我不确定进一步地对batch内的所有数据求整体均值是否有更多收益，或许还可能带来一些想不到的问题。

【English】
Sorry, I haven't conducted similar experiments before, so I can't say for sure. 
However, when using IFT to calculate the loss for each data point, it's akin to performing length normalization (dividing by loss_mask.sum(-1) accomplishes this). In my understanding, this normalization actually helps mitigate the influence of length distribution to some extent. 
So I'm not sure if there's more to gain from further calculating the overall mean for all data within the batch, and it might even introduce some unforeseen issues.

**Comment by peterjc123** [comment]
@Hua-rookie 确实，实验结果如你所说的，gsm8k的结果还是不太好，或许是我们数据集里面的短数据太多了

**Comment by b3y0nd** [comment]
> 【中文】 我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum
> 
> 【English】 We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum

这个库貌似不支持fp16和Bf16的运算，想问下作者训练时的参数设置是怎么样的？有什么办法解决这个问题吗

**Comment by peterjc123** [comment]
> > 【中文】 我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum
> > 【English】 We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum
> 
> 这个库貌似不支持fp16和Bf16的运算，想问下作者训练时的参数设置是怎么样的？有什么办法解决这个问题吗

我实现的时候logps是在float32这边计算的

**Comment by b3y0nd** [comment]
> > > 【中文】 我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum
> > > 【English】 We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum
> > 
> > 
> > 这个库貌似不支持fp16和Bf16的运算，想问下作者训练时的参数设置是怎么样的？有什么办法解决这个问题吗
> 
> 我实现的时候logps是在float32这边计算的

看来解决方案一致，我这边也是先转float32然后再计算cumsum_weight ：）

**Comment by Hua-rookie** [comment]
> > > > 【中文】 我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum
> > > > 【English】 We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum
> > > 
> > > 
> > > 这个库貌似不支持fp16和Bf16的运算，想问下作者训练时的参数设置是怎么样的？有什么办法解决这个问题吗
> > 
> > 
> > 我实现的时候logps是在float32这边计算的
> 
> 看来解决方案一致，我这边也是先转float32然后再计算cumsum_weight ：）

我们在实现的时候也是先用fp32。如果有其他用于计算强化学习cumsum的库，欢迎推荐~

**Comment by wuzhiyue111** [comment]
> > > > > 【中文】 我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum
> > > > > 【English】 We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum
> > > > 
> > > > 
> > > > 这个库貌似不支持fp16和Bf16的运算，想问下作者训练时的参数设置是怎么样的？有什么办法解决这个问题吗
> > > 
> > > 
> > > 我实现的时候logps是在float32这边计算的
> > 
> > 
> > 看来解决方案一致，我这边也是先转float32然后再计算cumsum_weight ：）
> 
> 我们在实现的时候也是先用fp32。如果有其他用于计算强化学习cumsum的库，欢迎推荐~

您好，我是第一次使用torch-discounted-cumsum这个包，我在pip install -r requirements.txt之后，执行
from torch_discounted_cumsum import discounted_cumsum_right, discounted_cumsum_left
报错
torch_discounted_cumsum Unknown CUDA arch (8.7) or GPU not supported

这个是什么原因呀

**Comment by Hua-rookie** [comment]
不好意思，我们没遇到过类似的情况，也不太清楚

**Comment by Hua-rookie** [comment]
> > > > > > 【中文】 我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum
> > > > > > 【English】 We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum
> > > > > 
> > > > > 
> > > > > 这个库貌似不支持fp16和Bf16的运算，想问下作者训练时的参数设置是怎么样的？有什么办法解决这个问题吗
> > > > 
> > > > 
> > > > 我实现的时候logps是在float32这边计算的
> > > 
> > > 
> > > 看来解决方案一致，我这边也是先转float32然后再计算cumsum_weight ：）
> > 
> > 
> > 我们在实现的时候也是先用fp32。如果有其他用于计算强化学习cumsum的库，欢迎推荐~
> 
> 您好，我是第一次使用torch-discounted-cumsum这个包，我在pip install -r requirements.txt之后，执行 from torch_discounted_cumsum import discounted_cumsum_right, discounted_cumsum_left 报错 torch_discounted_cumsum Unknown CUDA arch (8.7) or GPU not supported
> 
> 这个是什么原因呀

不好意思，我们没遇到过类似的情况，也不太清楚

**Comment by wuzhiyue111** [comment]
> > > > > > > 【中文】 我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum
> > > > > > > 【English】 We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum
> > > > > > 
> > > > > > 
> > > > > > 这个库貌似不支持fp16和Bf16的运算，想问下作者训练时的参数设置是怎么样的？有什么办法解决这个问题吗
> > > > > 
> > > > > 
> > > > > 我实现的时候logps是在float32这边计算的
> > > > 
> > > > 
> > > > 看来解决方案一致，我这边也是先转float32然后再计算cumsum_weight ：）
> > > 
> > > 
> > > 我们在实现的时候也是先用fp32。如果有其他用于计算强化学习cumsum的库，欢迎推荐~
> > 
> > 
> > 您好，我是第一次使用torch-discounted-cumsum这个包，我在pip install -r requirements.txt之后，执行 from torch_discounted_cumsum import discounted_cumsum_right, discounted_cumsum_left 报错 torch_discounted_cumsum Unknown CUDA arch (8.7) or GPU not supported
> > 这个是什么原因呀
> 
> 不好意思，我们没遇到过类似的情况，也不太清楚

可以问一下大佬的cuda版本是多少吗，用的啥卡吗，我H100是跑不起来的。。。。

**Comment by Hua-rookie** [comment]
> > > > > > > > 【中文】 我们推荐使用下面这个库参与 Dynamic Relation Propagation 的运算：https://github.com/toshas/torch-discounted-cumsum
> > > > > > > > 【English】 We recommend using the following repository for participating in Dynamic Relation Propagation calculations: https://github.com/toshas/torch-discounted-cumsum
> > > > > > > 
> > > > > > > 
> > > > > > > 这个库貌似不支持fp16和Bf16的运算，想问下作者训练时的参数设置是怎么样的？有什么办法解决这个问题吗
> > > > > > 
> > > > > > 
> > > > > > 我实现的时候logps是在float32这边计算的
> > > > > 
> > > > > 
> > > > > 看来解决方案一致，我这边也是先转float32然后再计算cumsum_weight ：）
> > > > 
> > > > 
> > > > 我们在实现的时候也是先用fp32。如果有其他用于计算强化学习cumsum的库，欢迎推荐~
> > > 
> > > 
> > > 您好，我是第一次使用torch-discounted-cumsum这个包，我在pip install -r requirements.txt之后，执行 from torch_discounted_cumsum import discounted_cumsum_right, discounted_cumsum_left 报错 torch_discounted_cumsum Unknown CUDA arch (8.7) or GPU not supported
> > > 这个是什么原因呀
> > 
> > 
> > 不好意思，我们没遇到过类似的情况，也不太清楚
> 
> 可以问一下大佬的cuda版本是多少吗，用的啥卡吗，我H100是跑不起来的。。。。

CUDA 12.2，A6000 和 A800 都跑过
